### PR TITLE
Vecsim coordinator

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR introduces coordinator modifications for VSS.
1. Introduces new private command `_REQIURED_FIELDS` between the coordinator and shard
2. Introduces additional sorting phase (if required) for sorting by vector score. This phase will be called if no SORTBY was given, or the SORTBY field is different than the vector field.

### CODE REVIEW

File: coord/src/module.c [hunk 1]
```diff
@@ -1180,6 +1558,53 @@ int LocalSearchCommandHandler(RedisModuleCtx *ctx, RedisModuleString **argv, int
...
+      if(req->requiredFields == NULL) {
+        req->requiredFields = array_new(const char*, 1);
+      }
+      req->requiredFields = array_append(req->requiredFields, ctx->sortby.sortKey);
+      // Sortkey is the first required key value to return
+      ctx->sortby.offset = 0;
+      offset++;
+    }
+
+    if(ctx->specialCaseType == SPECIAL_CASE_KNN) {
```
Comment (ashtul) [inline coord/src/module.c:1580]: `else if` or `switch`

**Review by ashtul — COMMENTED** [review]
Looks good. Few comments.

File: coord/src/module.c [hunk 2]
```diff
@@ -577,17 +768,133 @@ static void getReplyOffsets(const searchRequestCtx *ctx, searchReplyOffsets *off
...
+      rCtx->errorOccured = true;
+      // invalid result - usually means something is off with the response, and we should just
+      // quit this response
+      rCtx->cachedResult = res;
+      break;
+    } else {
+      rCtx->cachedResult = NULL;
+    }
+
+    scoredSearchResultWrapper* resWrapper = rm_malloc(sizeof(scoredSearchResultWrapper));
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply coord/src/module.c:870]: Good catch

File: coord/src/rmr/reply.h [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #define MR_REPLY_NIL 4
 #define MR_REPLY_STATUS 5
 #define MR_REPLY_ERROR 6
+#define MR_REPLY_DOUBLE 7
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply coord/src/rmr/reply.h]: reverted

File: src/aggregate/aggregate_exec.c [hunk 4]
```diff
@@ -33,6 +29,44 @@ typedef struct {
   const PLN_ArrangeStep *lastAstp;
 } cachedVars;
 
+static void reeval_key(RedisModuleCtx *outctx, const RSValue *key) {
+  RedisModuleString *rskey = NULL;
+  if (key) {
+    if(key->t == RSValue_Reference) {
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_exec.c:35]: Technically, the `RSValue_Dereference` should be done in a loop no?

File: src/aggregate/aggregate_exec.c [hunk 5]
```diff
@@ -33,6 +29,44 @@ typedef struct {
...
+      key = RSValue_Dereference(key);
+    }
+    switch (key->t) {
+      case RSValue_Number:
+        /* Serialize double - by prepending "%" to the number, so the coordinator/client can
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_exec.c:40]:
```suggestion
        /* Serialize double - by prepending "#" to the number, so the coordinator/client can
```

File: src/aggregate/aggregate_exec.c [hunk 6]
```diff
@@ -33,6 +29,44 @@ typedef struct {
...
+        break;
+    }
+    if (rskey) {
+      RedisModule_ReplyWithString(outctx, rskey);
+      RedisModule_FreeString(outctx, rskey);
+    } else {
+      RedisModule_ReplyWithNull(outctx);
+    }
+  } else {
+    RedisModule_ReplyWithNull(outctx);
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_exec.c:66]: nitpick:
I would have put if first so it will be more readable:
```
if (!key) {
    RedisModule_ReplyWithNull(outctx);
}
```

File: src/aggregate/aggregate_request.c [hunk 7]
```diff
@@ -114,42 +114,30 @@ static int parseCursorSettings(AREQ *req, ArgsCursor *ac, QueryError *status) {
...
-    const char *value = AC_GetStringNC(&paramsArgs, &value_len);
-    // FIXME: Validate param is [a-zA-Z][a-zA-z_\-:0-9]*
-    if (DICT_ERR == Param_DictAdd(params, param, value, value_len, status)) {
-      Param_DictFree(params);
-      return REDISMODULE_ERR;
-    }
+  int requiredFieldNum = AC_NumArgs(&args);
+  const char** requiredFields = array_new(const char*, requiredFieldNum);
+  for(size_t i=0; i < requiredFieldNum; i++) {
+    requiredFields = array_append(requiredFields, AC_GetStringNC(&args, NULL));
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_request.c:129]: You do not check input correctness here, you just continue to read until `requiredFieldNum` reached, do you check it elsewhere?

File: src/query_param.c [hunk 8]
```diff
@@ -195,3 +195,35 @@ int QueryParam_Resolve(Param *param, dict *params, QueryError *status) {
   }
   return -1;
 }
+
+int parseParams (dict **destParams, ArgsCursor *ac, QueryError *status) {
```
Comment (MeirShpilraien) [inline src/query_param.c:199]: Why this was moved here while `parseRequiredFields` was implemented elsewhere?
In general I am fine with moving code around but it should be done on a separate PR, here its hard to know if you made any changes to this function or just moved it...

File: src/aggregate/aggregate_request.c [hunk 9]
```diff
@@ -931,10 +924,9 @@ static ResultProcessor *getGroupRP(AREQ *req, PLN_GroupStep *gstp, ResultProcess
 static ResultProcessor *getVecSimRP(AREQ *req, AGGPlan *pln, ResultProcessor *up, QueryError *status) {
   RLookup *lk = AGPLN_GetLookup(pln, NULL, AGPLN_GETLOOKUP_LAST);
   char **scoreFields = req->ast.vecScoreFieldNames;
-  size_t len = array_len(scoreFields);
-  RLookupKey **keys = rm_calloc(len, sizeof(*keys));
-
-  for (size_t i = 0; i < len; i++) {
+  size_t nScoreFields = array_len(scoreFields);
+  RLookupKey **keys = rm_calloc(nScoreFields, sizeof(*keys));
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_request.c:928]: We can allocate it on the stack no? Or even use the array_on_stack you implemented on gears?

Comment (MeirShpilraien) [reply src/aggregate/aggregate_request.c:928]: Forget...

File: tests/pytests/test_vecsim.py [hunk 10]
```diff
@@ -207,7 +207,6 @@ def testCreate(env):
     # assertInfoField(env, 'idx5', 'attributes', info)
 
 def testCreateErrors(env):
-    env.skipOnCluster()
```
Comment (MeirShpilraien) [inline tests/pytests/test_vecsim.py:210]: Can you explain why you removed it?

Comment (MeirShpilraien) [reply tests/pytests/test_vecsim.py:210]: Ohh ok its the vector tests which was disable till now on cluster :+1:

File: tests/pytests/test_vecsim.py [hunk 11]
```diff
@@ -470,18 +471,31 @@ def test_hybrid_query_batches_mode_with_tags(env):
...
     execute_hybrid_query(env, '(@tags:{hybrid} @text:hello)=>[TOP_K 10 @v $vec_param]', query_data, 'tags').equal([0L])
-    expected_res_1 = [10L, '50', ['__v_score', '0', 'tags', 'hybrid'], '51', ['__v_score', '128', 'tags', 'hybrid'], '49', ['__v_score', '128', 'tags', 'hybrid'], '52', ['__v_score', '512', 'tags', 'hybrid'], '48', ['__v_score', '512', 'tags', 'hybrid'], '53', ['__v_score', '1152', 'tags', 'hybrid'], '47', ['__v_score', '1152', 'tags', 'hybrid'], '54', ['__v_score', '2048', 'tags', 'hybrid'], '46', ['__v_score', '2048', 'tags', 'hybrid'], '45', ['__v_score', '3200', 'tags', 'hybrid']]
+    if env.isCluster():
+        expected_res_1 = [10L, '50', ['__v_score', '0', 'tags', 'hybrid'], '49', ['__v_score', '128', 'tags', 'hybrid'], '51', ['__v_score', '128', 'tags', 'hybrid'], '48', ['__v_score', '512', 'tags', 'hybrid'], '52', ['__v_score', '512', 'tags', 'hybrid'], '47', ['__v_score', '1152', 'tags', 'hybrid'], '53', ['__v_score', '1152', 'tags', 'hybrid'], '46', ['__v_score', '2048', 'tags', 'hybrid'], '54', ['__v_score', '2048', 'tags', 'hybrid'], '45', ['__v_score', '3200', 'tags', 'hybrid']]
```
Comment (MeirShpilraien) [inline tests/pytests/test_vecsim.py:475]: Is the results on different in ordering? If so lets make the test more robust and resistance to replies order, otherwise lets add a comment why the results are different on cluster?

File: tests/pytests/test_vecsim.py [hunk 12]
```diff
@@ -470,18 +471,31 @@ def test_hybrid_query_batches_mode_with_tags(env):
...
+    if env.isCluster():
+        expected_res_2 = [10L, '50', ['__v_score', '0', 'tags', 'different, tag'], '40', ['__v_score', '12800', 'tags', 'different, tag'], '60', ['__v_score', '12800', 'tags', 'different, tag'], '30', ['__v_score', '51200', 'tags', 'different, tag'], '70', ['__v_score', '51200', 'tags', 'different, tag'], '20', ['__v_score', '115200', 'tags', 'different, tag'], '80', ['__v_score', '115200', 'tags', 'different, tag'], '10', ['__v_score', '204800', 'tags', 'different, tag'], '90', ['__v_score', '204800', 'tags', 'different, tag'], '100', ['__v_score', '320000', 'tags', 'different, tag']]
```
Comment (MeirShpilraien) [inline tests/pytests/test_vecsim.py:486]: Same comment as above

File: tests/pytests/test_vecsim.py [hunk 13]
```diff
@@ -470,18 +471,31 @@ def test_hybrid_query_batches_mode_with_tags(env):
...
+
     execute_hybrid_query(env, '(@tags:{different})=>[TOP_K 10 @v $vec_param]', query_data, 'tags').equal(expected_res_2)
-    expected_res_3 = [10L, '51', ['__v_score', '128', 'tags', 'hybrid'], '49', ['__v_score', '128', 'tags', 'hybrid'], '52', ['__v_score', '512', 'tags', 'hybrid'], '48', ['__v_score', '512', 'tags', 'hybrid'], '53', ['__v_score', '1152', 'tags', 'hybrid'], '47', ['__v_score', '1152', 'tags', 'hybrid'], '54', ['__v_score', '2048', 'tags', 'hybrid'], '46', ['__v_score', '2048', 'tags', 'hybrid'], '55', ['__v_score', '3200', 'tags', 'hybrid'], '45', ['__v_score', '3200', 'tags', 'hybrid']]
+    
+    if env.isCluster():
+        expected_res_3 = [10L, '49', ['__v_score', '128', 'tags', 'hybrid'], '51', ['__v_score', '128', 'tags', 'hybrid'], '48', ['__v_score', '512', 'tags', 'hybrid'], '52', ['__v_score', '512', 'tags', 'hybrid'], '47', ['__v_score', '1152', 'tags', 'hybrid'], '53', ['__v_score', '1152', 'tags', 'hybrid'], '46', ['__v_score', '2048', 'tags', 'hybrid'], '54', ['__v_score', '2048', 'tags', 'hybrid'], '45', ['__v_score', '3200', 'tags', 'hybrid'], '55', ['__v_score', '3200', 'tags', 'hybrid']]
```
Comment (MeirShpilraien) [inline tests/pytests/test_vecsim.py:493]: Same

File: coord/src/module.c [hunk 14]
```diff
@@ -1215,17 +1649,16 @@ int FlatSearchCommandHandler(RedisModuleBlockedClient *bc, RedisModuleString **a
...
+
+  if(req->specialCases) {
+    sendRequiredFields(req, &cmd);
   }
 
+
   struct MRCtx *mrctx = MR_CreateCtx(ctx, req);
   // we prefer the next level to be local - we will only approach nodes on our own shard
   // we also ask only masters to serve the request, to avoid duplications by random
-  MR_SetCoordinationStrategy(mrctx, MRCluster_FlatCoordination | MRCluster_MastersOnly);
```
Comment (MeirShpilraien) [inline coord/src/module.c:1228]: Why was this removed? IIRC this fixes a bug we just had of duplicate results. If its a wrong merge we need to make sure the other conflicts was not resolved incorrectly ...

File: coord/src/module.c [hunk 15]
```diff
@@ -657,12 +971,53 @@ static void processSearchReply(MRReply *arr, searchReducerCtx *rCtx, RedisModule
   }
 }
 
+/************************ Result post processing callbacks ********************/
+
+
+static void noOp(searchReducerCtx *rCtx){
```
Comment (MeirShpilraien) [inline coord/src/module.c:977]: Maybe change it to `noOpPostProcess`? Like `knnPostProcess`?

File: coord/src/module.c [hunk 16]
```diff
@@ -577,17 +769,139 @@ static void getReplyOffsets(const searchRequestCtx *ctx, searchReplyOffsets *off
...
+static void proccessKNNSearchReply(MRReply *arr, searchReducerCtx *rCtx, RedisModuleCtx *ctx) {
+  if (arr == NULL) {
+    return;
+  }
+  if (MRReply_Type(arr) == MR_REPLY_ERROR) {
+    rCtx->lastError = arr;
+    return;
+  }
+  if (MRReply_Type(arr) != MR_REPLY_ARRAY || MRReply_Length(arr) == 0) {
+    // Empty reply??
```
Comment (MeirShpilraien) [inline coord/src/module.c:840]: What will happened here?

File: src/aggregate/aggregate_exec.c (hunk 4 above)
Comment (DvirDukhan) [reply src/aggregate/aggregate_exec.c:35]: it is a loop, no?
```
static inline RSValue *RSValue_Dereference(const RSValue *v) {
  for (; v && v->t == RSValue_Reference; v = v->ref)
    ;
  return (RSValue *)v;
}
```

File: src/aggregate/aggregate_request.c (hunk 7 above)
Comment (DvirDukhan) [reply src/aggregate/aggregate_request.c:129]: `AC_GetVarArgs` checks about the var args array
added a check for NULL string
thanks

File: src/query_param.c (hunk 8 above)
Comment (DvirDukhan) [reply src/query_param.c:199]: I moved it since the coordinator needs to use it. The coordinator now includes `query_param.h`

File: src/query.c [hunk 17]
```diff
@@ -947,8 +947,6 @@ IndexIterator *Query_EvalNode(QueryEvalCtx *q, QueryNode *n) {
   return NULL;
 }
 
-QueryNode *RSQuery_ParseRaw(QueryParseCtx *);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/query.c:950]: it is on query.h

File: coord/src/module.c (hunk 14 above)
Comment (DvirDukhan) [reply coord/src/module.c:1228]: by accident

File: coord/src/module.c (hunk 16 above)
Comment (DvirDukhan) [reply coord/src/module.c:840]: empty reply? This is a copy from `processSearchReply` so it should be unified. let's discuss if needed

File: src/query.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/query.c:950]: Ohh you removed it :+1:

File: src/query_param.c (hunk 8 above)
Comment (MeirShpilraien) [reply src/query_param.c:199]: And nothing was changed? Just moved?

File: coord/src/module.c (hunk 16 above)
Comment (MeirShpilraien) [reply coord/src/module.c:840]: Let verify what will happened here? Its not setting an error no?

File: coord/src/module.c (hunk 14 above)
Comment (MeirShpilraien) [reply coord/src/module.c]: Did you check other merge conflict that did not removed important fixes accidentally?

File: src/aggregate/aggregate_request.c [hunk 18]
```diff
@@ -126,7 +126,13 @@ static int parseRequiredFields(AREQ *req, ArgsCursor *ac, QueryError *status){
   int requiredFieldNum = AC_NumArgs(&args);
   const char** requiredFields = array_new(const char*, requiredFieldNum);
   for(size_t i=0; i < requiredFieldNum; i++) {
-    requiredFields = array_append(requiredFields, AC_GetStringNC(&args, NULL));
+    const char *s = AC_GetStringNC(&args, NULL); {
+      if(!s) {
+        array_free(requiredFields);
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_request.c:131]: Don't you need to also free the strings that was already added to `requiredFields`?

Comment (MeirShpilraien) [reply src/aggregate/aggregate_request.c:131]: Also do we have test for this case?

File: src/aggregate/aggregate_request.c (hunk 7 above)
Comment (MeirShpilraien) [reply src/aggregate/aggregate_request.c]: Look at the last comment I posted.

File: src/aggregate/aggregate_request.c (hunk 18 above)
Comment (DvirDukhan) [reply src/aggregate/aggregate_request.c:131]: it is `const char *`, and a direct result of `AC_GetString` which is either `RedisModule_StringPtrLen` or `AC_CURRENT`. Anyway, it is not a copy

Comment (MeirShpilraien) [reply src/aggregate/aggregate_request.c:131]: If its not a copy then how do you know it will not be freed when you try to use it on `req->requiredFields`. Also what about a test that check this flow?

File: coord/src/module.c (hunk 16 above)
Comment (DvirDukhan) [reply coord/src/module.c:840]: it should not. let's assume you have only 1 data point on 3 shards. 2 shards will return an empty response. Will add a test

File: src/aggregate/aggregate_request.c [hunk 19]
```diff
@@ -124,6 +124,8 @@ static int parseRequiredFields(AREQ *req, ArgsCursor *ac, QueryError *status){
   }
 
   int requiredFieldNum = AC_NumArgs(&args);
+  // This array contains shallow copy of the required fields names. Those copies are to use only for lookup.
+  // If we need to use them in reply we should make a copy of those strings.
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_request.c:128]: :+1:

**Review by MeirShpilraien — APPROVED** [review]
:+1:

**Review by MeirShpilraien — DISMISSED** [review]

**Review by ashtul — APPROVED** [review]
